java/src/com/facebook/watchman/WatchmanClientImpl.java

```diff
@@ -255,11 +255,13 @@ public void call(Map<String, Object> message) throws Exception {
             .name(subscriptionId)
             .root(root)
             .build();
-        if (!subscriptions.containsKey(subscription)) {
+
+        Callback listener = subscriptions.get(subscription);
+        if (listener == null) {
```

Comment:
Yeah, you're right, the messaging handling is single threaded and it looks like it's not possible for this race condition to happen.